From fa046dece8cdc49dda3b3e469b7ff05231fe50cf Mon Sep 17 00:00:00 2001
From: Kurt Newman <kurt.newman@cpanel.net>
Date: Thu, 14 Jan 2016 15:31:16 -0600
Subject: [PATCH] Fix absolute/relative URI redirect in mod_rewrite

Bug 58854: This fixes a bug in Apache 2.4 where a specifying
a RewriteRule containing a relative path would not be properly
assigned to the redirect handler if it was a sub request.  The
result is that when /403.shtml is missing, the contents of
foo.txt were displayed even though it shouldn't be due to
the 'Require' directive preventing it.

For example:
    httpd.conf, add:
        ErrorDocument 403 /403.shtml

    .htaccess file, add:
        Require all denied
        RewriteEngine on
        RewriteRule .* foo.txt [L]

This patch should go away once Apache applies this to the
upstream version.

https://bz.apache.org/bugzilla/show_bug.cgi?id=58854
Signed-off-by: Kurt Newman <kurt.newman@cpanel.net>
---
 modules/mappers/mod_rewrite.c | 27 ++++++++++++++-------------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/modules/mappers/mod_rewrite.c b/modules/mappers/mod_rewrite.c
index 8156a7d..70981a0 100644
--- a/modules/mappers/mod_rewrite.c
+++ b/modules/mappers/mod_rewrite.c
@@ -5010,19 +5010,6 @@ static int hook_fixup(request_rec *r)
                 return HTTP_BAD_REQUEST;
             }
 
-            /* Check for deadlooping:
-             * At this point we KNOW that at least one rewriting
-             * rule was applied, but when the resulting URL is
-             * the same as the initial URL, we are not allowed to
-             * use the following internal redirection stuff because
-             * this would lead to a deadloop.
-             */
-            if (ofilename != NULL && strcmp(r->filename, ofilename) == 0) {
-                rewritelog((r, 1, dconf->directory, "initial URL equal rewritten"
-                            " URL: %s [IGNORING REWRITE]", r->filename));
-                return OK;
-            }
-
             tmpfilename = r->filename;
 
             /* if there is a valid base-URL then substitute
@@ -5083,6 +5070,20 @@ static int hook_fixup(request_rec *r)
                 }
             }
 
+            /* Check for deadlooping:
+             * At this point we KNOW that at least one rewriting
+             * rule was applied, but when the resulting URL is
+             * the same as the initial URL, we are not allowed to
+             * use the following internal redirection stuff because
+             * this would lead to a deadloop.
+             */
+            if (ofilename != NULL && strcmp(r->filename, ofilename) == 0) {
+                rewritelog((r, 1, dconf->directory, "initial URL equal rewritten"
+                            " URL: %s [IGNORING REWRITE]", r->filename));
+                return OK;
+            }
+
+
             /* now initiate the internal redirect */
             rewritelog((r, 1, dconf->directory, "internal redirect with %s "
                         "[INTERNAL REDIRECT]", r->filename));
-- 
2.6.4

